File: src/main/java/org/polyfrost/hytils/command/SilentRemoveCommand.java
/*
 * Hytils Reborn - Hypixel focused Quality of Life mod.
 * Copyright (C) 2020, 2021, 2022, 2023  Polyfrost, Sk1er LLC and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.polyfrost.hytils.command;

import dev.deftu.omnicore.api.client.chat.OmniClientChat;
import dev.deftu.textile.Text;
import dev.deftu.textile.minecraft.MCTextStyle;
import dev.deftu.textile.minecraft.TextColors;
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.Command;
import org.polyfrost.hytils.handlers.chat.modules.triggers.SilentRemoval;
import com.mojang.authlib.GameProfile;
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.Handler;
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.Param;

import java.util.Set;
import java.util.regex.Pattern;

@Command("silentremove")
public class SilentRemoveCommand {

    protected static final Pattern usernameRegex = Pattern.compile("\\w{1,16}");

    @Handler
    public void main() {
        OmniClientChat.displayChatMessage(Text.literal("Usage: /silentremove <player>").setStyle(MCTextStyle.color(TextColors.RED)));
    }

    @Handler(description = "Adds or removes a player from the silent list.")
    private void main(@Param("Player Name") GameProfile player) {
        String name = player.getName();
        if (!usernameRegex.matcher(name).matches()) {
            OmniClientChat.displayChatMessage(Text.literal("Invalid username.").setStyle(MCTextStyle.color(TextColors.RED)));
            return;
        }
        final Set<String> silentUsers = SilentRemoval.getSilentUsers();
        if (silentUsers.contains(name)) {
            silentUsers.remove(name);
            OmniClientChat.displayChatMessage(Text.literal("Removed ").setStyle(MCTextStyle.color(TextColors.GREEN)).append(Text.literal(name).setStyle(MCTextStyle.color(TextColors.YELLOW))).append(Text.literal(" from the silent removal list.").setStyle(MCTextStyle.color(TextColors.GREEN))));
            return;
        }

        silentUsers.add(name);
        OmniClientChat.displayChatMessage("&aAdded &e" + name + " &ato the removal queue.");
    }

    @Handler(description = "Clears the silent removal queue.")
    private void clear() {
        final Set<String> silentUsers = SilentRemoval.getSilentUsers();
        if (silentUsers.isEmpty()) {
            OmniClientChat.displayChatMessage(Text.literal("Silent removal list is already empty.").setStyle(MCTextStyle.color(TextColors.RED)));
            return;
        }

        silentUsers.clear();
        OmniClientChat.displayChatMessage(Text.literal("Cleared the silent removal list.").setStyle(MCTextStyle.color(TextColors.GREEN)));
    }
}

File: src/main/kotlin/org/polyfrost/hytils/client/commands/impl/SilentRemoveCommand.kt
package org.polyfrost.hytils.client.commands.impl

import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import net.fabricmc.fabric.api.client.command.v2.ClientCommands
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import org.polyfrost.hytils.client.commands.ClientCommand
import org.polyfrost.hytils.client.features.chat.handlers.modules.triggers.SilentRemoval
import org.polyfrost.hytils.client.data.providers.LanguageData
import org.polyfrost.hytils.client.utils.ChatUtils
import java.util.Locale

object SilentRemoveCommand : ClientCommand {
    override fun getCommand(): LiteralArgumentBuilder<FabricClientCommandSource> = ClientCommands.literal("silentremove")
        .then(ClientCommands.literal("add")
            .then(ClientCommands.argument("username", StringArgumentType.string())
                .executes { context ->
                    val username = context.getArgument("username", String::class.java)
                    executeAdd(username.lowercase(Locale.ROOT))
                    Command.SINGLE_SUCCESS
                }
            )
        )
        .then(ClientCommands.literal("remove")
            .then(ClientCommands.argument("username", StringArgumentType.string())
                .executes { context ->
                    val username = context.getArgument("username", String::class.java)
                    executeRemove(username.lowercase(Locale.ROOT))
                    Command.SINGLE_SUCCESS
                }
            )
        )
        .then(ClientCommands.literal("list")
            .executes {
                executeList()
                Command.SINGLE_SUCCESS
            }
        )
        .then(ClientCommands.literal("clear")
            .executes {
                executeClear()
                Command.SINGLE_SUCCESS
            }
        )

    override fun getAliases() = listOf("sremove")

    private fun executeAdd(username: String) {
        if (!username.matches(LanguageData.VALID_USERNAME)) {
            ChatUtils.displayMessage(Component.literal("Invalid username!").withStyle(ChatFormatting.RED))
            return
        }

        if (!SilentRemoval.removalQueue.contains(username)) {
            SilentRemoval.removalQueue.add(username)
            ChatUtils.displayMessage(
                Component.literal("Added ").withStyle(ChatFormatting.GREEN)
                    .append(Component.literal(username).withStyle(ChatFormatting.YELLOW))
                    .append(Component.literal(" to the silent removal queue.").withStyle(ChatFormatting.GREEN))
            )
        } else {
            ChatUtils.displayMessage(
                Component.literal("$username is already in the silent removal queue!").withStyle(ChatFormatting.RED)
            )
        }
    }

    private fun executeRemove(username: String) {
        if (!username.matches(LanguageData.VALID_USERNAME)) {
            ChatUtils.displayMessage(Component.literal("Invalid username!").withStyle(ChatFormatting.RED))
            return
        }

        if (SilentRemoval.removalQueue.contains(username)) {
            SilentRemoval.removalQueue.remove(username)
            ChatUtils.displayMessage(
                Component.literal("Removed ").withStyle(ChatFormatting.GREEN)
                    .append(Component.literal(username).withStyle(ChatFormatting.YELLOW))
                    .append(Component.literal(" from the silent removal queue.").withStyle(ChatFormatting.GREEN))
            )
        } else {
            ChatUtils.displayMessage(
                Component.literal("$username is not in the silent removal queue!").withStyle(ChatFormatting.RED)
            )
        }
    }

    private fun executeList() {
        if (SilentRemoval.removalQueue.isEmpty()) {
            ChatUtils.displayMessage(
                Component.literal("The silent removal queue is empty.").withStyle(ChatFormatting.YELLOW)
            )
        } else {
            ChatUtils.displayMessage(
                Component.literal("Silent removal queue: ").withStyle(ChatFormatting.GREEN)
                    .apply {
                        SilentRemoval.removalQueue.forEachIndexed { index, username ->
                            append(Component.literal(username).withStyle(ChatFormatting.YELLOW))
                            if (index < SilentRemoval.removalQueue.size - 1) {
                                append(Component.literal(", ").withStyle(ChatFormatting.GREEN))
                            }
                        }
                    }
            )
        }
    }

    private fun executeClear() {
        if (SilentRemoval.removalQueue.isEmpty()) {
            ChatUtils.displayMessage(
                Component.literal("The silent removal queue is already empty.").withStyle(ChatFormatting.YELLOW)
            )
        } else {
            SilentRemoval.removalQueue.clear()
            ChatUtils.displayMessage(
                Component.literal("Cleared the silent removal queue.").withStyle(ChatFormatting.GREEN)
            )
        }
    }
}
